services/httpd/pprof.go

```
@@ -68,60 +62,173 @@ type prof struct {
-		// seconds to capture the profile for.
-		profile.Debug, _ = strconv.ParseInt(r.FormValue("seconds"), 10, 64)
-		if profile.Debug <= 0 {
-			profile.Debug = 30
+	// In the following two blocks, we check if the request should include cpu
+	// profiles and a trace log.
+	//
+	// Since the submitted form can contain multiple version of a variable like:
+	//
+	//   http://localhost:8086?cpu=1s&cpu=30s&trace=3s&cpu=5s
+	//
+	// the question arises: which value should we use?  We choose to use the LAST
+	// value supplied.
+	//
+	// This is an edge case but if for some reason, for example, a url is
+	// programatically built and multiple values are supplied, this will do what
+	// is expected.
+	//
+
+	// last() returns either the last item from a slice of strings or an empty
+	// string if the supplied slice is empty or nill.
+	last := func(s []string) string {
+		if len(s) == 0 {
+			return ""
 		}
-		allProfs = append([]*prof{profile}, allProfs...) // CPU profile first.
+		return s[len(s)-1]
 	}
 
-	var (
-		resp bytes.Buffer // Temporary buffer for entire archive.
-		buf  bytes.Buffer // Temporary buffer for each profile/query result.
-	)
+	// if trace exsits as a form value, add it to the profiles slice with the
+	// decoded duration.
+	//
+	// Requests for a trace should look like:
+	//
+	//  ?trace=10s
+	//
+	if vals, exists := r.Form["trace"]; exists {
+		// parse the duration encoded in the last "trace" value supplied.
+		duration, err := time.ParseDuration(last(vals))
+
+		// If we can't parse the duration or if the users supplies a negative
+		// number, then set it to a reasonable default of 10s.
+		//
+		// This allows for stuff like ?trace=true -- which isn't recommened but
+		// makes the code do something reasonable.
+		//
+		// If someone wants to supply a trace duration of 0s I guess that's their
+		// perogative but lets ensure that it is positive.
+		if defaultDuration := time.Second * 10; err != nil || duration < 0 {
```

If the user supplies a negative or zero duration, it should return an error. They might have a bug in a script that requests profiles or traces and not be collecting what they expected to collect. It should error on that so they see the bug instead of masking it.

---

makes sense.

---

fixed.